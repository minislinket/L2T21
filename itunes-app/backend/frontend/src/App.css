.header {
  margin-top: 16px;
  border: 4px solid #a0a0a0;
  border-radius: 5px;
}

.search-card {
  background-color: #cff4fc;
}
