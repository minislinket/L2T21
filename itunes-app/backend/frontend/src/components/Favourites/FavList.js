import { useEffect, useState } from "react";
import { Card, Container, ListGroup, Row } from "react-bootstrap";
import { AudiobookResult } from "../Home/SearchResults/AudiobookResult";
import { EbookResult } from "../Home/SearchResults/EbookResult";
import { MovieResult } from "../Home/SearchResults/MovieResult";
import { MusicResult } from "../Home/SearchResults/MusicResult";
import { PodcastResult } from "../Home/SearchResults/PodcastResult";
import { SoftwareResult } from "../Home/SearchResults/SoftwareResult";
import { TVShowResult } from "../Home/SearchResults/TVShowResult";
export function FavList(props) {
  /**
   * @type {array}
   * favouritesData is used to store the data received from the fetch API
   */
  const [favouritesData, setFavouritesData] = useState([]);
  /**
   * @type {object}
   * favComponents is used to store the components to be rendered within FavList
   */
  const [favComponents, setFavComponents] = useState(null);
  /**
   * @type {boolean}
   * loading is used to indicate to the user that the application is still
   * waiting to receive data from the API
   */
  const [loading, setLoading] = useState(false);
  /**
   * @type {object}
   * error is used to indicate to the user if there has been an error trying to
   * retrieve data from the API
   */
  const [error, setError] = useState(null);
  /**
   * @type {boolean}
   * nothing is used to indicate whether or not the user has saved any media to favourites
   */
  const [nothing, setNothing] = useState(false);

  // On initial load of the Component, call the function to load user favourites
  useEffect(() => {
    loadFavourites();
  }, []);

  // This function uses the fetch API to obtain the details of the user's favourited media.
  // We use the IDs of the favourited media to fetch the rest of the details
  function loadFavourites() {
    // Array of IDs to lookup
    let idLookup = [];
    // This is a check if we're unit testing this component, as IDs are saved in the local storage of the browser
    if (props.testIDs !== undefined) {
      idLookup = props.testIDs;
    } else {
      // Get the IDs from the local storage of the browser
      idLookup = localStorage.getItem("favourites");
    }
    fetch(
      //Instead of hassling with stringifying arrays into a URL, I use URLSearchParams to set the paramaters for the call
      // URL would look like: /favourites?idArray=%5B1498844775%2C1108257204%2C1490222697%2C287000993%2C1497290440%2C811377230%2C448142450%5D
      "/favourites?" + new URLSearchParams({ idArray: idLookup }).toString()
    )
      .then((response) => {
        // Get the response from the API call
        if (response.ok) {
          return response.json();
        }
        throw response;
      })
      .then((data) => {
        setFavouritesData(data); // Set the data to a usable state
      })
      .catch((error) => {
        console.error("Error fetching data: ", error); // Handle any error in fetching the data
        setError(error);
      })
      .finally(() => {
        setLoading(false); // We're no longer waiting for any response
      });
  }

  // Everytime favouritesData changes, render the components to display the data
  useEffect(() => {
    // If there are any favourited media, render the components to display them and remove the component that displays nothing
    if (favouritesData.resultCount > 0) {
      setNothing(false);
      setFavComponents(favouritesData.results.map(checkMediaKind));
    } else {
      setNothing(true);
    }
  }, [favouritesData]);

  // Depending on the media type/kind, return the respective Component to render
  function checkMediaKind(result, index, arr) {
    let mediaType = "";
    let key = "";
    // If the result has a kind value, set the media type to that value
    // If the result does not have a kind value, it has a wrapperType we can set the mediaType as
    if (result.kind !== undefined) {
      mediaType = result.kind;
      key = mediaType + index;
    } else {
      mediaType = result.wrapperType;
      key = mediaType + index;
    }
    // Switch case to return a Component respective to its mediaType
    switch (mediaType) {
      case "audiobook":
        return (
          <AudiobookResult
            key={key}
            index={index}
            result={result}
            onDelete={loadFavourites}
          />
        );
      case "ebook":
        return (
          <EbookResult
            key={key}
            index={index}
            result={result}
            onDelete={loadFavourites}
          />
        );
      case "feature-movie":
        return (
          <MovieResult
            key={key}
            index={index}
            result={result}
            onDelete={loadFavourites}
          />
        );
      case "song":
        return (
          <MusicResult
            key={key}
            index={index}
            result={result}
            onDelete={loadFavourites}
          />
        );
      case "podcast":
        return (
          <PodcastResult
            key={key}
            index={index}
            result={result}
            onDelete={loadFavourites}
          />
        );
      case "software":
        return (
          <SoftwareResult
            key={key}
            index={index}
            result={result}
            onDelete={loadFavourites}
          />
        );
      case "collection":
        return (
          <TVShowResult
            key={key}
            index={index}
            result={result}
            onDelete={loadFavourites}
          />
        );
    }
  }

  // If loading is set to true when fetching the data, display a loading Card
  function LoadingCard() {
    return (
      <Container style={{ marginTop: "15px" }}>
        <Card border="warning">
          <Card.Body>
            <Card.Title>Loading Page...</Card.Title>
            <Card.Text>Please wait while we load your page...</Card.Text>
          </Card.Body>
        </Card>
      </Container>
    );
  }
  // If the error state is truthy after trying to fetch the data, display an
  // error Card
  function ErrorCard() {
    return (
      <Container style={{ marginTop: "15px" }}>
        <Card border="danger">
          <Card.Body>
            <Card.Title>Error</Card.Title>
            <Card.Text>"Error fetching data"</Card.Text>
          </Card.Body>
        </Card>
      </Container>
    );
  }

  // If nothing is set to true, display a Card saying there are no favourites
  function NoFavouritesCard() {
    return (
      <Container style={{ marginTop: "15px" }}>
        <Card border="info">
          <Card.Body>
            <Card.Title>Nothing Here Yet!</Card.Title>
            <Card.Text>
              No Favourites added yet. Use this site to search the iTunes API
              for your favourite media!
            </Card.Text>
          </Card.Body>
        </Card>
      </Container>
    );
  }

  // Conditional rendering to display correct information
  return (
    <Container>
      <Row>
        <ListGroup>
          <ListGroup.Item variant="info">
            {loading && <LoadingCard />}
            {error && <ErrorCard />}
            {nothing && <NoFavouritesCard />}
            {!loading && !error && !nothing && favComponents}
          </ListGroup.Item>
        </ListGroup>
      </Row>
    </Container>
  );
}
