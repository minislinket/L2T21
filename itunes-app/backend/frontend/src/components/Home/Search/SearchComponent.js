import { useRef, useState } from "react";
import { Button, Card, Dropdown, Form } from "react-bootstrap";

// Search Box Component
function SearchComponent(props) {
  // We use the useRef hooks below to help store the values the user inputs
  const searchRef = useRef("");
  /**
   * @type {string}
   * This state is used to change the text on the dropdown button
   */
  const [dropdownButtonText, setDropdownButtonText] =
    useState("Select Media Type");

  // With this function we collect the user inputs and search using the iTunes API
  function searchItunesHandler(event) {
    // Collect the user input values using the useRef hooks
    const searchTerm = searchRef.current.value;
    const mediaType = dropdownButtonText.replace(" ", ""); // Sanitise mediaType value for use

    // Simple validation disallowing empty strings and checking for dropdown buton text
    if (
      !searchTerm ||
      onlySpaces(searchTerm) ||
      dropdownButtonText === "Select Media Type"
    ) {
      alert("Search field must not be empty | Select a media type");
    } else {
      const searchObj = {
        term: searchTerm,
        media: mediaType,
      };
      // Send the object back to App.js to be used in searching the iTunes API
      props.onSubmit(searchObj);
    }
  }

  // This function returns true if a string with only spaces is passed i.e. " "
  function onlySpaces(str) {
    return str.trim().length === 0;
  }
  // Array of different media the user can search for
  const mediaArray = [
    "Any",
    "Movie",
    "Podcast",
    "Music",
    "Audiobook",
    "TV Show",
    "Software",
    "eBook",
  ];

  // This function changes the dropdown button text to the dropdown item selected
  function changeDropdownButtonText(e) {
    setDropdownButtonText(e.target.text);
  }

  return (
    <Card className="search-card">
      <Card.Body>
        <Form>
          <Form.Group className="mb-3" controlId="formBasicSearch">
            <Form.Label>Search</Form.Label>
            <Form.Control
              type="text"
              placeholder="Enter a search term"
              ref={searchRef}
            />
          </Form.Group>
          <Dropdown className="mb-3">
            <Dropdown.Toggle variant="success" id="dropdown-basic">
              {dropdownButtonText}
            </Dropdown.Toggle>

            <Dropdown.Menu>
              {mediaArray.map((media, index) => (
                <Dropdown.Item
                  href="#"
                  key={"media" + index}
                  onClick={changeDropdownButtonText}
                >
                  {media}
                </Dropdown.Item>
              ))}
            </Dropdown.Menu>
          </Dropdown>
          <Button variant="primary" type="button" onClick={searchItunesHandler}>
            Submit
          </Button>
        </Form>
      </Card.Body>
    </Card>
  );
}

export default SearchComponent;
