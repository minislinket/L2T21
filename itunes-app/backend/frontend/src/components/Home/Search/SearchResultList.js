import { useEffect, useState } from "react";
import { ListGroup } from "react-bootstrap";
import { AudiobookResult } from "../SearchResults/AudiobookResult";
import { EbookResult } from "../SearchResults/EbookResult";
import { MovieResult } from "../SearchResults/MovieResult";
import { MusicResult } from "../SearchResults/MusicResult";
import { PodcastResult } from "../SearchResults/PodcastResult";
import { SoftwareResult } from "../SearchResults/SoftwareResult";
import { TVShowResult } from "../SearchResults/TVShowResult";

// The Component that displays the search results
export function SearchResultList(props) {
  /**
   * @type {object}
   * resultsComponents is used to store the components to be rendered within SearchResultList
   */
  const [resultsComponents, setResultsComponents] = useState(null);

// Everytime searchResultData changes, render the components to display the data
  useEffect(() => {
    // If there is any media matching the search term, render the components to display them
    if (props.searchResultData.resultCount > 0) {
      setResultsComponents(props.searchResultData.results.map(checkMediaKind));
    }
  }, [props.searchResultData]);

  // Depending on the media type/kind, return the respective Component to render
  function checkMediaKind(result, index, arr) {
    let mediaType = "";
    let key = "";
    // If the result has a "kind" value, set the media type to that value
    // If the result does not have a "kind" value, it has a "wrapperType" we can set the mediaType as
    if (result.kind !== undefined) {
      mediaType = result.kind;
      key = mediaType + index;
    } else {
      mediaType = result.wrapperType;
      key = mediaType + index;
    }
    // Switch case to return a Component respective to its mediaType
    switch (mediaType) {
      case "audiobook":
        return <AudiobookResult key={key} index={index} result={result} />;
      case "ebook":
        return <EbookResult key={key} index={index} result={result} />;
      case "feature-movie":
        return <MovieResult key={key} index={index} result={result} />;
      case "song":
        return <MusicResult key={key} index={index} result={result} />;
      case "podcast":
        return <PodcastResult key={key} index={index} result={result} />;
      case "software":
        return <SoftwareResult key={key} index={index} result={result} />;
      case "collection":
        return <TVShowResult key={key} index={index} result={result} />;
    }
  }

  return (
    <ListGroup >
      <ListGroup.Item variant="success">{resultsComponents}</ListGroup.Item>
    </ListGroup>
  );
}
