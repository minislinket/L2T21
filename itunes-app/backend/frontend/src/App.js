import "bootstrap/dist/css/bootstrap.min.css"; // Import some styling from Bootstrap
import { FavList } from "./components/Favourites/FavList";

import { useEffect } from "react";
import { BrowserRouter, Route, Routes } from "react-router-dom";
import "./App.css";
import Header from "./components/Header/Header";
import HomePage from "./components/Home/HomePage";

function App() {
  // On initial load, we initialise the "inFavourites" value in session storage if there isnt any there 
  useEffect(() => {
    if (sessionStorage.getItem("inFavourites") === null) { //if there is no session data
      sessionStorage.setItem("inFavourites", false); //set value as inFavourites and set value to false
    }
  }, []);

  return (
    <div>
      <BrowserRouter>
        <div>
          <Header />
          <hr></hr>
          <Routes>
            <Route exact={true} path="/" element={<HomePage />}></Route>
            <Route path="/fav" element={<FavList />}></Route>
          </Routes>
        </div>
      </BrowserRouter>
    </div>
  );
}

export default App;
