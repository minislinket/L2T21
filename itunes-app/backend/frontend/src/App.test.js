import renderer from "react-test-renderer";
import App from "./App";
import { FavList } from "./components/Favourites/FavList";
import { SearchResultList } from "./components/Home/Search/SearchResultList";

// Test data to use for SearchResultList
const searchResultData = [
  {
    wrapperType: "track",
    kind: "feature-movie",
    trackId: 1108257204,
    artistName: "Michael Showalter",
    trackName: "Hello, My Name Is Doris",
    trackCensoredName: "Hello, My Name Is Doris",
    trackViewUrl:
      "https://itunes.apple.com/za/movie/hello-my-name-is-doris/id1108257204?uo=4",
    previewUrl:
      "https://video-ssl.itunes.apple.com/itunes-assets/Video127/v4/12/7c/34/127c340a-f411-4194-6b58-39fd26f6ec0b/mzvf_7747801430432270584.640x480.h264lc.U.p.m4v",
    artworkUrl30:
      "https://is3-ssl.mzstatic.com/image/thumb/Video50/v4/80/03/60/800360e0-7a81-159f-7226-0f31ba6100d2/HELLO_MY_NAME_IS_DORIS_2016_TH_INTL_Logo-WW-artwork.lsr/30x30bb.jpg",
    artworkUrl60:
      "https://is3-ssl.mzstatic.com/image/thumb/Video50/v4/80/03/60/800360e0-7a81-159f-7226-0f31ba6100d2/HELLO_MY_NAME_IS_DORIS_2016_TH_INTL_Logo-WW-artwork.lsr/60x60bb.jpg",
    artworkUrl100:
      "https://is3-ssl.mzstatic.com/image/thumb/Video50/v4/80/03/60/800360e0-7a81-159f-7226-0f31ba6100d2/HELLO_MY_NAME_IS_DORIS_2016_TH_INTL_Logo-WW-artwork.lsr/100x100bb.jpg",
    collectionPrice: 99.99,
    trackPrice: 99.99,
    releaseDate: "2016-03-11T08:00:00Z",
    collectionExplicitness: "notExplicit",
    trackExplicitness: "notExplicit",
    trackTimeMillis: 5396062,
    country: "ZAF",
    currency: "ZAR",
    primaryGenreName: "Comedy",
    raw: {
      wrapperType: "track",
      kind: "feature-movie",
      trackId: 1108257204,
      artistName: "Michael Showalter",
      trackName: "Hello, My Name Is Doris",
      trackCensoredName: "Hello, My Name Is Doris",
      trackViewUrl:
        "https://itunes.apple.com/za/movie/hello-my-name-is-doris/id1108257204?uo=4",
      previewUrl:
        "https://video-ssl.itunes.apple.com/itunes-assets/Video127/v4/12/7c/34/127c340a-f411-4194-6b58-39fd26f6ec0b/mzvf_7747801430432270584.640x480.h264lc.U.p.m4v",
      artworkUrl30:
        "https://is3-ssl.mzstatic.com/image/thumb/Video50/v4/80/03/60/800360e0-7a81-159f-7226-0f31ba6100d2/HELLO_MY_NAME_IS_DORIS_2016_TH_INTL_Logo-WW-artwork.lsr/30x30bb.jpg",
      artworkUrl60:
        "https://is3-ssl.mzstatic.com/image/thumb/Video50/v4/80/03/60/800360e0-7a81-159f-7226-0f31ba6100d2/HELLO_MY_NAME_IS_DORIS_2016_TH_INTL_Logo-WW-artwork.lsr/60x60bb.jpg",
      artworkUrl100:
        "https://is3-ssl.mzstatic.com/image/thumb/Video50/v4/80/03/60/800360e0-7a81-159f-7226-0f31ba6100d2/HELLO_MY_NAME_IS_DORIS_2016_TH_INTL_Logo-WW-artwork.lsr/100x100bb.jpg",
      collectionPrice: 99.99,
      trackPrice: 99.99,
      collectionHdPrice: 129.99,
      trackHdPrice: 129.99,
      releaseDate: "2016-03-11T08:00:00Z",
      collectionExplicitness: "notExplicit",
      trackExplicitness: "notExplicit",
      trackTimeMillis: 5396062,
      country: "ZAF",
      currency: "ZAR",
      primaryGenreName: "Comedy",
      contentAdvisoryRating: "10-12 PG",
      shortDescription:
        "After a lifetime of being overlooked and ignored, a woman of a certain age finds her world turned",
      longDescription:
        "After a lifetime of being overlooked and ignored, a woman of a certain age finds her world turned upside down by a handsome new co-worker and a self-help seminar that inspires her to take a chance on love in Hello, My Name is Doris, a witty and compassionate late-life coming-of-age-story.",
      hasITunesExtras: true,
    },
    releaseDate: "2016-03-11T08:00:00Z",
    trackCensoredName: "Hello, My Name Is Doris",
    trackExplicitness: "notExplicit",
    trackId: 1108257204,
    trackName: "Hello, My Name Is Doris",
    trackPrice: 99.99,
    trackTimeMillis: 5396062,
    trackViewUrl:
      "https://itunes.apple.com/za/movie/hello-my-name-is-doris/id1108257204?uo=4",
    wrapperType: "track",
  },
  {
    wrapperType: "track",
    kind: "podcast",
    collectionId: 811377230,
    trackId: 811377230,
    artistName: "CGP Grey & Brady Haran",
    collectionName: "Hello Internet",
    trackName: "Hello Internet",
    collectionCensoredName: "Hello Internet",
    trackCensoredName: "Hello Internet",
    collectionViewUrl:
      "https://podcasts.apple.com/za/podcast/hello-internet/id811377230?uo=4",
    trackViewUrl:
      "https://podcasts.apple.com/za/podcast/hello-internet/id811377230?uo=4",
    artworkUrl30:
      "https://is5-ssl.mzstatic.com/image/thumb/Podcasts6/v4/19/33/fe/1933fe85-cd86-2191-8187-d725ca7359bf/mza_8038397602264410223.png/30x30bb.jpg",
    artworkUrl60:
      "https://is5-ssl.mzstatic.com/image/thumb/Podcasts6/v4/19/33/fe/1933fe85-cd86-2191-8187-d725ca7359bf/mza_8038397602264410223.png/60x60bb.jpg",
    artworkUrl100:
      "https://is5-ssl.mzstatic.com/image/thumb/Podcasts6/v4/19/33/fe/1933fe85-cd86-2191-8187-d725ca7359bf/mza_8038397602264410223.png/100x100bb.jpg",
    collectionPrice: 0,
    trackPrice: 0,
    releaseDate: "2020-02-28T10:13:00Z",
    collectionExplicitness: "cleaned",
    trackExplicitness: "cleaned",
    trackCount: 100,
    country: "ZAF",
    currency: "ZAR",
    primaryGenreName: "Education",
    raw: {
      wrapperType: "track",
      kind: "podcast",
      collectionId: 811377230,
      trackId: 811377230,
      artistName: "CGP Grey & Brady Haran",
      collectionName: "Hello Internet",
      trackName: "Hello Internet",
      collectionCensoredName: "Hello Internet",
      trackCensoredName: "Hello Internet",
      collectionViewUrl:
        "https://podcasts.apple.com/za/podcast/hello-internet/id811377230?uo=4",
      feedUrl: "http://www.hellointernet.fm/podcast?format=rss",
      trackViewUrl:
        "https://podcasts.apple.com/za/podcast/hello-internet/id811377230?uo=4",
      artworkUrl30:
        "https://is5-ssl.mzstatic.com/image/thumb/Podcasts6/v4/19/33/fe/1933fe85-cd86-2191-8187-d725ca7359bf/mza_8038397602264410223.png/30x30bb.jpg",
      artworkUrl60:
        "https://is5-ssl.mzstatic.com/image/thumb/Podcasts6/v4/19/33/fe/1933fe85-cd86-2191-8187-d725ca7359bf/mza_8038397602264410223.png/60x60bb.jpg",
      artworkUrl100:
        "https://is5-ssl.mzstatic.com/image/thumb/Podcasts6/v4/19/33/fe/1933fe85-cd86-2191-8187-d725ca7359bf/mza_8038397602264410223.png/100x100bb.jpg",
      collectionPrice: 0,
      trackPrice: 0,
      trackRentalPrice: 0,
      collectionHdPrice: 0,
      trackHdPrice: 0,
      trackHdRentalPrice: 0,
      releaseDate: "2020-02-28T10:13:00Z",
      collectionExplicitness: "cleaned",
      trackExplicitness: "cleaned",
      trackCount: 100,
      country: "ZAF",
      currency: "ZAR",
      primaryGenreName: "Education",
      contentAdvisoryRating: "Clean",
      artworkUrl600:
        "https://is5-ssl.mzstatic.com/image/thumb/Podcasts6/v4/19/33/fe/1933fe85-cd86-2191-8187-d725ca7359bf/mza_8038397602264410223.png/600x600bb.jpg",
      genreIds: ["1304", "26"],
      genres: ["Education", "Podcasts"],
    },
    releaseDate: "2020-02-28T10:13:00Z",
    trackCensoredName: "Hello Internet",
    trackCount: 100,
    trackExplicitness: "cleaned",
    trackId: 811377230,
    trackName: "Hello Internet",
    trackPrice: 0,
    trackViewUrl:
      "https://podcasts.apple.com/za/podcast/hello-internet/id811377230?uo=4",
    wrapperType: "track",
  },
];

// Test data to use for FavList
const testIDs = [1108257204, 811377230]

test("snapshot test of SearchResultList", () => {
  const tree = renderer
    .create(<SearchResultList searchResultData={searchResultData} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("snapshot test of SearchResultList", async () => {
  const tree = renderer
    .create(<FavList testIDs={testIDs} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
