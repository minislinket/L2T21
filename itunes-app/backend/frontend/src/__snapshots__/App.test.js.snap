// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test of SearchResultList 1`] = `
<div
  className="list-group"
  onKeyDown={[Function]}
>
  <div
    className="list-group-item list-group-item-success"
    data-rr-ui-event-key={null}
    onClick={[Function]}
  />
</div>
`;

exports[`snapshot test of SearchResultList 2`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="list-group"
      onKeyDown={[Function]}
    >
      <div
        className="list-group-item list-group-item-info"
        data-rr-ui-event-key={null}
        onClick={[Function]}
      />
    </div>
  </div>
</div>
`;
